### PR TITLE
[Small]  fix #976, explicitly define `NestedTensorSpec` and `NestedTensor`.

### LINKED ISSUE #976: Create Type Hint for "Nest(ed Structure) of Tensors"
# Motivation

Type hints helps improve readability a lot. Nest of Tensors is a category of types that are used extensively in alf, but we do not have a way to correctly annotate them. This includes Nest of `TensorSpec` as well.

# Proposed Solution

It is proposed that we can define a customized type hint that looks like `alf.Nest[Tensor]`.

# Tasks

- [ ] Define the custom type hint `alf.Nest[xxx]`
- [ ] Annotate nest of tensors and nest of tensor specs in critical places

### PR DESCRIPTION
**Pull request by breakds**
## Motivation 

Currently both are defined as `Any`. Explicitly define them can slightly improve readability. Fix #976.

## Note

Also, `NestedTensorSpec` has been moved to `tensor_spec.py`, and we can now refer to it as `alf.NestedTensorSpec`.

### CODE REVIEW

File: alf/networks/normalizing_flow_networks.py [hunk 1]
```diff
@@ -44,12 +44,11 @@ class NormalizingFlowNetwork(Network):
...
     """
 
-    def __init__(
-            self,
-            input_tensor_spec: alf.tensor_specs.TensorSpec,
-            conditional_input_tensor_spec: alf.nest.NestedTensorSpec = None,
-            use_transform_cache: bool = True,
-            name: str = "NormalizingFlowNetwork"):
+    def __init__(self,
+                 input_tensor_spec: alf.tensor_specs.TensorSpec,
```
Comment (hnyu) [inline alf/networks/normalizing_flow_networks.py]: alf.TensorSpec

File: alf/networks/normalizing_flow_networks.py [hunk 2]
```diff
@@ -212,22 +211,21 @@ class RealNVPNetwork(NormalizingFlowNetwork):
...
-            activation: Callable = torch.tanh,
-            transform_scale_nonlinear: Callable = partial(
-                clipped_exp, clip_value_min=-10, clip_value_max=2),
-            sub_dim: int = None,
-            mask_mode: str = "contiguous",
-            num_layers: int = 2,
-            use_transform_cache: bool = True,
-            name: str = "RealNVPNetwork"):
+    def __init__(self,
+                 input_tensor_spec: alf.tensor_specs.TensorSpec,
```
Comment (hnyu) [inline alf/networks/normalizing_flow_networks.py]: alf.TensorSpec

File: alf/networks/normalizing_flow_networks.py [hunk 3]
```diff
@@ -360,7 +358,7 @@ def _make_invertible_transform(self, conditional_inputs=None):
 def _prepare_conditional_flow_inputs(
         xy_spec: alf.tensor_specs.TensorSpec,
```
Comment (hnyu) [inline alf/networks/normalizing_flow_networks.py]: alf.TensorSpec

File: alf/networks/normalizing_flow_networks.py [hunk 4]
```diff
@@ -423,15 +421,14 @@ class _RealNVPTransform(td.Transform):
...
-            self,
-            input_tensor_spec: alf.tensor_specs.TensorSpec,
-            scale_trans_net: EncodingNetwork,
-            mask: torch.Tensor,
-            conditional_input_tensor_spec: alf.nest.NestedTensorSpec = None,
-            z: alf.nest.NestedTensor = None,
-            cache_size: int = 1,
-            scale_nonlinear: Callable = torch.exp):
+    def __init__(self,
+                 input_tensor_spec: alf.tensor_specs.TensorSpec,
```
Comment (hnyu) [inline alf/networks/normalizing_flow_networks.py]: alf.TensorSpec

File: alf/tensor_specs.py [hunk 5]
```diff
@@ -380,3 +381,7 @@ def numpy_sample(self, outer_dims=None):
                 high=self._maximum + 1,
                 size=shape,
                 dtype=torch_dtype_to_str(self._dtype))
+
+
+NestedTensorSpec = Union[TensorSpec, List['NestedTensorSpec'], Tuple[(
+)], Tuple['NestedTensorSpec', ...], Dict[str, 'NestedTensorSpec']]
```
Comment (hnyu) [inline alf/tensor_specs.py]: What's the difference between having this "..." with not having it?

File: alf/networks/normalizing_flow_networks.py (hunk 1 above)
Comment (breakds) [reply alf/networks/normalizing_flow_networks.py]: Updated.

File: alf/networks/normalizing_flow_networks.py (hunk 2 above)
Comment (breakds) [reply alf/networks/normalizing_flow_networks.py]: Updated.

File: alf/networks/normalizing_flow_networks.py (hunk 3 above)
Comment (breakds) [reply alf/networks/normalizing_flow_networks.py]: Updated.

File: alf/networks/normalizing_flow_networks.py (hunk 4 above)
Comment (breakds) [reply alf/networks/normalizing_flow_networks.py]: Updated.

File: alf/tensor_specs.py (hunk 5 above)
Comment (breakds) [reply alf/tensor_specs.py]: `Tuple['NestedTensorSpec']` stands for a tuple of a single element. It is hard to cover the exact case of tuple of variant length of `NestedTensorSpec`. Added documentation for readability here (and also for `NestedTensor`).

**Review by breakds — COMMENTED** [review]
Also disabled `yapf` for those two definitions, as it really hurts readability.

**Review by hnyu — APPROVED** [review]
